### PR TITLE
Fixed #31516 -- Improved naming of migrations with multiple operations.

### PR DESCRIPTION
**Pull request by manav014**
[ticket-31516](https://code.djangoproject.com/ticket/31516)

Supersedes [#12799  ](https://github.com/django/django/pull/12799)

### CODE REVIEW

File: django/db/migrations/migration.py [hunk 1]
```diff
@@ -184,15 +183,24 @@ def suggest_name(self):
...
+        name = fragments[0]
+        left_out = len(fragments) - 1
+        for fragment in fragments[1:]:
+            new_name = name + f"_{fragment}"
+            if len(new_name) > 52:
+                break
+            name = new_name
+            left_out -= 1
+        if(left_out > 0):
+            name = name + f"_+{left_out}"
```
Comment (felixxm) [inline django/db/migrations/migration.py:203]: We shouldn't use `+` in filenames, moreover the number of additional operations doesn't have much value.

`_and_more` sounds like a good marker (see [comment](https://github.com/django/django/pull/12799#issuecomment-636694424)).

**Review by felixxm — COMMENTED** [review]
@manav014 Thanks for this patch :+1: It looks like you missed my [comments](https://github.com/django/django/pull/12799#issuecomment-747898443).

File: django/db/migrations/migration.py (hunk 1 above)
Comment (felixxm) [reply django/db/migrations/migration.py]: > What about `_16_more`(16 is the number of operations)?

I don't see much value in this format,  you still need to open the file to find out more.

**Comment by felixxm** [comment]
@manav014 My [comments](https://github.com/django/django/pull/12799#issuecomment-747898443) are not solved. Please check them before marking the patch as ready for review.

File: django/db/migrations/migration.py [hunk 2]
```diff
@@ -184,15 +183,24 @@ def suggest_name(self):
...
-        if name is None:
-            name = 'initial' if self.initial else 'auto_%s' % get_migration_name_timestamp()
+        if self.initial:
+            return "initial"
+
+        raw_fragments = (op.migration_name_fragment for op in self.operations)
+        fragments = [name for name in raw_fragments if name]
+
+        if not fragments:
+            return f"auto_{get_migration_name_timestamp()}"
```
Comment (felixxm) [inline django/db/migrations/migration.py:193]: f-strings shouldn't contain function calls. This guideline is from [Python coding style](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#python-style).

**Comment by khu5h1** [comment]
@felixxm Thank you for your suggestions.

> IMO we should use "auto" names if any operation doesn't have a suggested name, because it can be confusing to return the same name for different set of operations
I have added a check for the operations with no suggested name.

> we could add some mark to inform that a migration contains other operations but their names were truncated (ellipsis is of course a bad idea):
I have added `_and_more` in the name of the migration file for this.

File: django/db/migrations/migration.py [hunk 3]
```diff
@@ -184,15 +183,25 @@ def suggest_name(self):
...
+        name = fragments[0]
+        left_out = len(fragments) - 1
+        for fragment in fragments[1:]:
+            new_name = name + f"_{fragment}"
+            if len(new_name) > 52:
+                break
+            name = new_name
+            left_out -= 1
+        if(left_out > 0):
+            name = name + "_and_more"
```
Comment (felixxm) [inline django/db/migrations/migration.py:204]: `left_out` is not necessary anymore:
```suggestion
        name = fragments[0]
        for fragment in fragments[1:]:
            new_name = f'{name}_{fragment}'
            if len(new_name) > 52:
                name = f'{name}_and_more'
                break
            name = new_name
```

File: tests/migrations/test_autodetector.py [hunk 4]
```diff
@@ -2663,16 +2663,83 @@ class Migration(migrations.Migration):
...
+        migration = Migration("some_migration", "test_app")
+        self.assertEqual(
+            migration.suggest_name(),
+            "person1_person2_person3_person4_person5_person6_and_more",
+        )
+
+    def test_operation_with_no_suggested_name(self):
+        class Migration(migrations.Migration):
+            operations = [
+                migrations.CreateModel("Person1", fields=[]),
```
Comment (felixxm) [inline tests/migrations/test_autodetector.py:2725]: Use single-quotes in the new code.

File: tests/migrations/test_autodetector.py [hunk 5]
```diff
@@ -2663,16 +2663,83 @@ class Migration(migrations.Migration):
...
+                migrations.RunSQL('SELECT 1 FROM person;'),
+            ]
+        migration = Migration("some_migration", "test_app")
+        self.assertIs(migration.suggest_name().startswith('auto_'), True)
+
+    def test_single_create_model_long_name(self):
+        """
+        Check that the first migration_name_fragment values isn't cut when
```
Comment (felixxm) [inline tests/migrations/test_autodetector.py:2733]: To add more explanation, docstrings should state the expected behavior and omit prefixes like "Check that" since all tests check things. This guideline is from [Python coding style](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#python-style).

**Comment by manav014** [comment]
@felixxm Thank you for your suggestions.

>IMO we should use "auto" names if any operation doesn't have a suggested name, because it can be confusing to return the same name for different set of operations

I have added a check for the operations with no suggested name.

>we could add some mark to inform that a migration contains other operations but their names were truncated (ellipsis is of course a bad idea):

I have added `_and_more` in the name of the migration file for this.

File: django/db/migrations/migration.py [hunk 6]
```diff
@@ -184,15 +183,25 @@ def suggest_name(self):
...
-            name = 'initial' if self.initial else 'auto_%s' % get_migration_name_timestamp()
+        if self.initial:
+            return "initial"
+
+        raw_fragments = (op.migration_name_fragment for op in self.operations)
+        fragments = [name for name in raw_fragments]
+
+        migration_name_timestamp = get_migration_name_timestamp()
+        if not fragments or None in fragments:
+            return f"auto_{migration_name_timestamp}"
```
Comment (felixxm) [inline django/db/migrations/migration.py:194]: We can move `get_migration_name_timestamp()` call to `if`.

File: django/db/migrations/migration.py [hunk 7]
```diff
@@ -184,15 +183,25 @@ def suggest_name(self):
...
-        if name is None:
-            name = 'initial' if self.initial else 'auto_%s' % get_migration_name_timestamp()
+        if self.initial:
+            return "initial"
+
+        raw_fragments = (op.migration_name_fragment for op in self.operations)
+        fragments = [name for name in raw_fragments]
+
+        migration_name_timestamp = get_migration_name_timestamp()
+        if not fragments or None in fragments:
```
Comment (felixxm) [inline django/db/migrations/migration.py:193]: We should probably check for empty strings as well.

**Review by felixxm — COMMENTED** [review]
@manav014 Thanks for updates :+1: I left more comments.

@adamchainz Do you want to take a look?

File: django/db/migrations/migration.py (hunk 3 above)
Comment (manav014) [reply django/db/migrations/migration.py:204]: I used it as a way to check for the need of `_and_more`. So in order to check for that I feel this is required. Is there any other optimized way to check for that which I am missing?

Comment (felixxm) [reply django/db/migrations/migration.py:204]: My comment contains a suggestion.

File: django/db/migrations/migration.py [hunk 8]
```diff
@@ -184,15 +183,22 @@ def suggest_name(self):
...
-            name = self.operations[0].migration_name_fragment
-        elif (
-            len(self.operations) > 1 and
-            all(isinstance(o, operations.CreateModel) for o in self.operations)
-        ):
-            name = '_'.join(sorted(o.migration_name_fragment for o in self.operations))
-        if name is None:
-            name = 'initial' if self.initial else 'auto_%s' % get_migration_name_timestamp()
+        if self.initial:
+            return 'initial'
```
Comment (felixxm) [inline django/db/migrations/migration.py:187]: I've moved the initial migrations changes to a separate PR, see #14116.

File: tests/migrations/test_autodetector.py [hunk 9]
```diff
@@ -2663,16 +2663,83 @@ class Migration(migrations.Migration):
...
 
         migration = Migration('0001_initial', 'test_app')
         suggest_name = migration.suggest_name()
         self.assertIs(suggest_name.startswith('auto_'), True)
 
+    def test_no_fragments(self):
+        class Migration(migrations.Migration):
+            operations = [migrations.RunSQL("SELECT 1")]
+        migration = Migration('some_migration', 'test_app')
+        self.assertTrue(migration.suggest_name().startswith('auto_'))
```
Comment (felixxm) [inline tests/migrations/test_autodetector.py:2686]: This is redundant with `test_none_name()`.

File: django/db/migrations/migration.py (hunk 8 above)
Comment (manav014) [reply django/db/migrations/migration.py:187]: Acknowledged.
May you help me with the changes that I shall commit in [this PR](https://github.com/django/django/pull/14109)? Or Shall I just remove Conflicts(the `if condition` for the initial name)?

**Review by felixxm — COMMENTED** [review]
@manav014 Thanks for updates :+1: 

I rebased from `main`, pushed edits, and moved tests for migrations with no operations to a separate commit.

**Comment by adamchainz** [comment]
Thanks for picking this up and merging @manav014 @felixxm !
